src/main/scala/io/iohk/ethereum/blockchain/sync/SyncStateScheduler.scala

```diff
@@ -0,0 +1,450 @@
+  }
+
+  private def processActiveResponse(
+      state: SchedulerState,
+      activeRequest: StateNodeRequest,
+      response: SyncResponse
+  ): Either[ResponseProcessingError, SchedulerState] = {
+    activeRequest.requestType match {
+      case _: CodeRequest => Right(state.commit(activeRequest.copy(resolvedData = Some(response.data))))
+      case requestType: NodeRequest =>
+        for {
+          mptNode <- Try(response.data.toArray.toMptNode).toEither.left.map(_ => CannotDecodeMptNode)
+          possibleChildRequests <- createPossibleChildRequests(mptNode, activeRequest, requestType)
+        } yield {
+          val childWithoutAlreadyKnown = possibleChildRequests.filterNot(req => isRequestAlreadyKnown(state, req))
+          if (childWithoutAlreadyKnown.isEmpty && activeRequest.dependencies == 0) {
+            state.commit(activeRequest.copy(resolvedData = Some(response.data)))
+          } else {
+            state.resolveRequest(activeRequest, response.data, childWithoutAlreadyKnown)
+          }
+        }
+    }
+  }
+
+  def processResponse(
+      state: SchedulerState,
+      response: SyncResponse
+  ): Either[ResponseProcessingError, SchedulerState] = {
+    for {
+      activeRequest <- isRequestAlreadyKnownOrResolved(state, response)
+      newState <- processActiveResponse(state, activeRequest, response)
+    } yield newState
+
+  }
+  // scalastyle:off method.length
+  private def createPossibleChildRequests(
+      mptNode: MptNode,
+      parentRequest: StateNodeRequest,
+      requestType: NodeRequest
+  ): Either[NotAccountLeafNode.type, Seq[StateNodeRequest]] = mptNode match {
+    case n: LeafNode =>
+      requestType match {
+        case SyncStateScheduler.StateNode =>
+          Account(n.value).toEither.left.map(_ => NotAccountLeafNode).map { account =>
+            // We are scheduling both storage trie and code requests with highest priority to be sure that leaf nodes completed
+            // as fast as possible
+            val evmRequests = if (account.codeHash != emptyCodeHash) {
+              Seq(StateNodeRequest(account.codeHash, None, Code, Seq(parentRequest.nodeHash), maxMptTrieDepth, 0))
+            } else {
+              Seq()
+            }
+
+            val storageRequests = if (account.storageRoot != emptyStateRootHash) {
+              Seq(
+                StateNodeRequest(
+                  account.storageRoot,
+                  None,
+                  StorageNode,
+                  Seq(parentRequest.nodeHash),
+                  maxMptTrieDepth,
+                  0
+                )
+              )
+            } else {
+              Seq()
+            }
+
+            evmRequests ++ storageRequests
+          }
+
+        case SyncStateScheduler.StorageNode =>
+          Right(Seq())
+      }
+
+    case n: BranchNode =>
+      Right(n.children.collect { case HashNode(childHash) =>
+        StateNodeRequest(
+          ByteString.fromArrayUnsafe(childHash),
```

Comment:
I assume the `ByteString.fromArrayUnsafe` is here for performance reasons, right?

Reply:
Yup, it can be used when we known that provided array won't be mutated. (as in this case)